core/services/workflows/artifacts/store.go

```diff
@@ -0,0 +1,330 @@
+		return math.MaxUint32
+	}
+	return uint32(n)
+}
+
+// FetcherFunc is an abstraction for fetching the contents stored at a URL.
+type FetcherFunc func(ctx context.Context, url string, maxBytesSize uint32) ([]byte, error)
+
+type ArtifactConfig struct {
+	MaxConfigSize  uint64
+	MaxSecretsSize uint64
+	MaxBinarySize  uint64
+}
+
+// By default, if type is unknown, the largest artifact size is 26.4KB.  Configure the artifact size
+// via the ArtifactConfig to override this default.
+const defaultMaxArtifactSizeBytes = 26.4 * utils.KB
+
+func (cfg *ArtifactConfig) ApplyDefaults() {
+	if cfg.MaxConfigSize == 0 {
+		cfg.MaxConfigSize = defaultMaxArtifactSizeBytes
+	}
+	if cfg.MaxSecretsSize == 0 {
+		cfg.MaxSecretsSize = defaultMaxArtifactSizeBytes
+	}
+	if cfg.MaxBinarySize == 0 {
+		cfg.MaxBinarySize = defaultMaxArtifactSizeBytes
+	}
+}
+
+// logCustMsg emits a custom message to the external sink and logs an error if that fails.
+func logCustMsg(ctx context.Context, cma custmsg.MessageEmitter, msg string, log logger.Logger) {
+	err := cma.Emit(ctx, msg)
+	if err != nil {
+		log.Helper(1).Errorf("failed to send custom message with msg: %s, err: %v", msg, err)
+	}
+}
+
+var defaultSecretsFreshnessDuration = 24 * time.Hour
+
+func WithMaxArtifactSize(cfg ArtifactConfig) func(*Store) {
+	return func(a *Store) {
+		a.limits = &cfg
+	}
+}
+
+type SerialisedModuleStore interface {
+	StoreModule(workflowID string, binaryID string, module []byte) error
+	GetModulePath(workflowID string) (string, bool, error)
+	GetBinaryID(workflowID string) (string, bool, error)
+	DeleteModule(workflowID string) error
+}
+
+type Store struct {
+	lggr logger.Logger
+
+	// limits sets max artifact sizes to fetch when handling events
+	limits *ArtifactConfig
+
+	orm WorkflowRegistryDS
+
+	// fetchFn is a function that fetches the contents of a URL with a limit on the size of the response.
+	fetchFn FetcherFunc
+
+	lastFetchedAtMap         *lastFetchedAtMap
+	clock                    clockwork.Clock
+	secretsFreshnessDuration time.Duration
+
+	encryptionKey workflowkey.Key
+
+	emitter custmsg.MessageEmitter
+}
+
+func NewStore(lggr logger.Logger, orm WorkflowRegistryDS, fetchFn FetcherFunc, clock clockwork.Clock, encryptionKey workflowkey.Key,
+	emitter custmsg.MessageEmitter, opts ...func(*Store)) *Store {
+	limits := &ArtifactConfig{}
+	limits.ApplyDefaults()
+
+	artifactsStore := &Store{
+		lggr:                     lggr,
+		orm:                      orm,
+		fetchFn:                  fetchFn,
+		lastFetchedAtMap:         newLastFetchedAtMap(),
+		clock:                    clock,
+		limits:                   limits,
+		secretsFreshnessDuration: defaultSecretsFreshnessDuration,
+		encryptionKey:            encryptionKey,
+		emitter:                  emitter,
+	}
+
+	for _, o := range opts {
+		o(artifactsStore)
+	}
+
+	return artifactsStore
+}
+
+// FetchWorkflowArtifacts fetches the workflow spec and config from a cache or the specified URLs if the artifacts have not
```

Comment:
I find it weird that a Store is doing fetching.

To me it would make more sense if this function was more something like `GetWorkflowArtifacts`, which would return if there is a cached artifact. If not fetching would still be handled by the Syncer handler.